File: 2023/day20/php/solution.php
<?php
/**
 * Day 20: Pulse Propagation - Module communication simulation.
 */

/**
 * Parse module configuration from input.
 */
function parseInput(string $filename): array {
    $modules = [];
    $lines = explode("\n", trim(file_get_contents($filename)));

    foreach ($lines as $line) {
        list($namePart, $destPart) = explode(' -> ', $line);
        $destinations = array_map('trim', explode(',', $destPart));

        if ($namePart === 'broadcaster') {
            $modules['broadcaster'] = [
                'type' => 'broadcaster',
                'destinations' => $destinations
            ];
        } elseif ($namePart[0] === '%') {
            $name = substr($namePart, 1);
            $modules[$name] = [
                'type' => 'flip-flop',
                'destinations' => $destinations,
                'state' => false
            ];
        } elseif ($namePart[0] === '&') {
            $name = substr($namePart, 1);
            $modules[$name] = [
                'type' => 'conjunction',
                'destinations' => $destinations,
                'memory' => []
            ];
        }
    }

    // Initialize conjunction memory for all inputs
    foreach ($modules as $name => $module) {
        foreach ($module['destinations'] as $dest) {
            if (isset($modules[$dest]) && $modules[$dest]['type'] === 'conjunction') {
                $modules[$dest]['memory'][$name] = false;
            }
        }
    }

    return $modules;
}

/**
 * Reset all module states.
 */
function resetModules(array &$modules): void {
    foreach ($modules as &$module) {
        if ($module['type'] === 'flip-flop') {
            $module['state'] = false;
        } elseif ($module['type'] === 'conjunction') {
            foreach ($module['memory'] as $key => $val) {
                $module['memory'][$key] = false;
            }
        }
    }
}

/**
 * Simulate a single button press.
 * Returns [low_count, high_count, high_senders_set]
 */
function simulateButtonPress(array &$modules, array $watchNodes = []): array {
    $lowCount = 0;
    $highCount = 0;
    $highSenders = [];

    // Queue: [source, destination, pulse] where pulse is true for high, false for low
    $queue = [['button', 'broadcaster', false]];

    while (!empty($queue)) {
        $item = array_shift($queue);
        list($source, $dest, $pulse) = $item;

        if ($pulse) {
            $highCount++;
        } else {
            $lowCount++;
        }

        // Track if watched nodes send high pulses
        if (!empty($watchNodes) && in_array($source, $watchNodes, true) && $pulse) {
            $highSenders[$source] = true;
        }

        if (!isset($modules[$dest])) {
            continue;
        }

        $module = &$modules[$dest];

        if ($module['type'] === 'broadcaster') {
            foreach ($module['destinations'] as $nextDest) {
                $queue[] = [$dest, $nextDest, $pulse];
            }
        } elseif ($module['type'] === 'flip-flop') {
            if (!$pulse) {  // Only react to low pulses
                $module['state'] = !$module['state'];
                foreach ($module['destinations'] as $nextDest) {
                    $queue[] = [$dest, $nextDest, $module['state']];
                }
            }
        } elseif ($module['type'] === 'conjunction') {
            $module['memory'][$source] = $pulse;
            // Send low if all inputs are high, otherwise send high
            $allHigh = true;
            foreach ($module['memory'] as $val) {
                if (!$val) {
                    $allHigh = false;
                    break;
                }
            }
            $output = !$allHigh;
            foreach ($module['destinations'] as $nextDest) {
                $queue[] = [$dest, $nextDest, $output];
            }
        }
    }

    return [$lowCount, $highCount, array_keys($highSenders)];
}

/**
 * Part 1: Count pulses after 1000 button presses.
 */
function part1(array $modules): int {
    resetModules($modules);

    $totalLow = 0;
    $totalHigh = 0;

    for ($i = 0; $i < 1000; $i++) {
        list($low, $high, $_) = simulateButtonPress($modules);
        $totalLow += $low;
        $totalHigh += $high;
    }

    return $totalLow * $totalHigh;
}

/**
 * GCD using GMP for large numbers.
 */
function gmpGcd($a, $b) {
    return gmp_gcd($a, $b);
}

/**
 * LCM using GMP for large numbers.
 */
function gmpLcm($a, $b) {
    return gmp_div_q(gmp_mul($a, $b), gmpGcd($a, $b));
}

/**
 * Part 2: Find minimum button presses for rx to receive a low pulse.
 */
function part2(array $modules): string {
    resetModules($modules);

    // Find the module that feeds into rx
    $rxInput = null;
    foreach ($modules as $name => $module) {
        if (in_array('rx', $module['destinations'], true)) {
            $rxInput = $name;
            break;
        }
    }

    if ($rxInput === null) {
        return "0";
    }

    // Find all modules that feed into rx_input
    $watchNodes = array_keys($modules[$rxInput]['memory']);
    $cycleLengths = [];

    $buttonPress = 0;
    while (count($cycleLengths) < count($watchNodes)) {
        $buttonPress++;
        list($_, $_, $highSenders) = simulateButtonPress($modules, $watchNodes);

        foreach ($highSenders as $node) {
            if (!isset($cycleLengths[$node])) {
                $cycleLengths[$node] = $buttonPress;
            }
        }
    }

    // LCM of all cycle lengths using GMP
    $result = gmp_init(1);
    foreach ($cycleLengths as $length) {
        $result = gmpLcm($result, gmp_init($length));
    }

    return gmp_strval($result);
}

// Main execution
$inputFile = __DIR__ . '/../input.txt';
$modules = parseInput($inputFile);
echo "Part 1: " . part1($modules) . "\n";

// Re-parse for part 2 (fresh state)
$modules = parseInput($inputFile);
echo "Part 2: " . part2($modules) . "\n";

File: 2023/day20/clojure/solution.clj
#!/usr/bin/env clojure -M

;; Day 20: Pulse Propagation - Module communication simulation

(require '[clojure.string :as str])

(defn parse-line [line]
  "Parse a single module configuration line."
  (let [[name-part dest-part] (str/split line #" -> ")
        destinations (mapv str/trim (str/split dest-part #","))]
    (cond
      (= name-part "broadcaster")
      ["broadcaster" {:type :broadcaster :destinations destinations}]

      (str/starts-with? name-part "%")
      [(subs name-part 1) {:type :flip-flop :destinations destinations :state false}]

      (str/starts-with? name-part "&")
      [(subs name-part 1) {:type :conjunction :destinations destinations :memory {}}]

      :else
      [name-part {:type :output :destinations destinations}])))

(defn parse-input [filename]
  "Parse module configuration from input file."
  (let [lines (str/split-lines (str/trim (slurp filename)))
        modules (into {} (map parse-line lines))]
    ;; Initialize conjunction memory for all inputs
    (reduce (fn [mods [name module]]
              (reduce (fn [m dest]
                        (if (and (contains? m dest)
                                 (= :conjunction (get-in m [dest :type])))
                          (assoc-in m [dest :memory name] false)
                          m))
                      mods
                      (:destinations module)))
            modules
            modules)))

(defn process-pulse [modules source dest pulse watch-nodes]
  "Process a single pulse and return [updated-modules new-pulses high-sender]."
  (let [high-sender (when (and (contains? watch-nodes source) pulse) source)]
    (if-not (contains? modules dest)
      [modules [] high-sender]
      (let [module (get modules dest)]
        (case (:type module)
          :broadcaster
          [modules
           (mapv #(vector dest % pulse) (:destinations module))
           high-sender]

          :flip-flop
          (if pulse
            ;; High pulse - ignore
            [modules [] high-sender]
            ;; Low pulse - toggle state
            (let [new-state (not (:state module))
                  updated-modules (assoc-in modules [dest :state] new-state)]
              [updated-modules
               (mapv #(vector dest % new-state) (:destinations module))
               high-sender]))

          :conjunction
          (let [updated-memory (assoc (:memory module) source pulse)
                updated-modules (assoc-in modules [dest :memory] updated-memory)
                output (not (every? true? (vals updated-memory)))]
            [updated-modules
             (mapv #(vector dest % output) (:destinations module))
             high-sender])

          ;; Default (output or unknown)
          [modules [] high-sender])))))

(defn simulate-button-press [modules watch-nodes]
  "Simulate a single button press.
   Returns [updated-modules low-count high-count high-senders]."
  (loop [mods modules
         queue (conj clojure.lang.PersistentQueue/EMPTY ["button" "broadcaster" false])
         low-count 0
         high-count 0
         high-senders #{}]
    (if (empty? queue)
      [mods low-count high-count high-senders]
      (let [[source dest pulse] (peek queue)
            remaining (pop queue)
            [new-low new-high] (if pulse [low-count (inc high-count)] [(inc low-count) high-count])
            [updated-mods new-pulses high-sender] (process-pulse mods source dest pulse watch-nodes)
            updated-senders (if high-sender (conj high-senders high-sender) high-senders)]
        (recur updated-mods
               (into remaining new-pulses)
               new-low
               new-high
               updated-senders)))))

(defn reset-state [modules]
  "Reset all module states to initial values."
  (reduce (fn [mods [name module]]
            (case (:type module)
              :flip-flop (assoc-in mods [name :state] false)
              :conjunction (assoc-in mods [name :memory]
                                     (into {} (map (fn [[k _]] [k false]) (:memory module))))
              mods))
          modules
          modules))

(defn part1 [modules]
  "Part 1: Count pulses after 1000 button presses."
  (loop [mods (reset-state modules)
         presses 0
         total-low 0
         total-high 0]
    (if (>= presses 1000)
      (* total-low total-high)
      (let [[new-mods low high _] (simulate-button-press mods #{})]
        (recur new-mods (inc presses) (+ total-low low) (+ total-high high))))))

(defn find-rx-input [modules]
  "Find the module that feeds into rx."
  (first (for [[name module] modules
               :when (some #(= % "rx") (:destinations module))]
           name)))

(defn gcd [a b]
  (if (zero? b) a (recur b (mod a b))))

(defn lcm [a b]
  (/ (* a b) (gcd a b)))

(defn part2 [modules]
  "Part 2: Find minimum button presses for rx to receive a low pulse.
   rx receives from a conjunction. For it to send low, all its inputs
   must have sent high. Find the cycle length for each input and compute LCM."
  (let [rx-input (find-rx-input modules)]
    (if (nil? rx-input)
      0
      (let [watch-nodes (set (keys (get-in modules [rx-input :memory])))]
        (loop [mods (reset-state modules)
               button-press 0
               cycle-lengths {}]
          (if (= (count cycle-lengths) (count watch-nodes))
            (reduce lcm 1 (vals cycle-lengths))
            (let [press (inc button-press)
                  [new-mods _ _ high-senders] (simulate-button-press mods watch-nodes)
                  new-cycles (reduce (fn [cycles node]
                                       (if (contains? cycles node)
                                         cycles
                                         (assoc cycles node press)))
                                     cycle-lengths
                                     high-senders)]
              (recur new-mods press new-cycles))))))))

(defn -main []
  (let [input-file (str (-> *file* java.io.File. .getParent) "/../input.txt")
        modules (parse-input input-file)]
    (println "Part 1:" (part1 modules))
    (println "Part 2:" (part2 (parse-input input-file)))))

(-main)
